### PR TITLE
Add tests

### PR DESCRIPTION
**Pull request by Ralender**
this is based on https://github.com/triSYCL/sycl/pull/111.

### CODE REVIEW

**Comment by keryell** [comment]
Rename `kernel_arguement_forwarding.cpp` → `kernel_argument_forwarding.cpp`

**Review by keryell — CHANGES_REQUESTED** [review]
Rename kernel_arguement_forwarding.cpp → kernel_argument_forwarding.cpp
Also I wonder whether there is a need for a rebase. The changes seem huge...

**Comment by Ralender** [comment]
addressed comments.

File: sycl/test/on-device/xocc/simple_tests/matmul.cpp [hunk 1]
```diff
@@ -0,0 +1,138 @@
...
+        B_rw[{i, j}] = gen_random();
+      }
+
+    printf("A:\n");
+    print(A_rw);
+    printf("B:\n");
+    print(B_rw);
+      matmul(G_rw, A_rw, B_rw);
+      printf("Gold:\n");
+      print(G_rw);\
```
Comment (keryell) [inline sycl/test/on-device/xocc/simple_tests/matmul.cpp]: Strange

File: sycl/test/on-device/xocc/simple_tests/matmul.cpp [hunk 2]
```diff
@@ -0,0 +1,138 @@
...
+  for (int r = 0; r < std::min(10ul, A_r.get_range()[1]); r++) {
+    printf("   %s ", "…");
+  }
+  printf("⋱\n\n");
+}
+
+void verify(cl::sycl::accessor<int, 2, cl::sycl::access::mode::read, cl::sycl::access::target::host_buffer> gold_r,
+            cl::sycl::accessor<int, 2, cl::sycl::access::mode::read, cl::sycl::access::target::host_buffer> output_r) {
+  for (size_t i = 0; i < (size_t)output_r.get_range()[0]; i++) {
+    for (size_t j = 0; j < (size_t)output_r.get_range()[1]; j++) {
```
Comment (keryell) [inline sycl/test/on-device/xocc/simple_tests/matmul.cpp]: No tautological cast needed here https://www.khronos.org/registry/SYCL/specs/sycl-2020/html/sycl-2020.html#range-class

File: sycl/test/on-device/xocc/simple_tests/matmul.cpp [hunk 3]
```diff
@@ -0,0 +1,138 @@
...
+  }
+
+  // Submitting command group(work) to queue
+  Queue.submit([&](cl::sycl::handler &cgh) {
+    // Getting write only access to the buffer on a device
+    auto C_w = C.get_access<cl::sycl::access::mode::read_write>(cgh);
+    auto A_r = A.get_access<cl::sycl::access::mode::read>(cgh);
+    auto B_r = B.get_access<cl::sycl::access::mode::read>(cgh);
+    // Executing kernel
+    cgh.single_task<class S>([=]() {
```
Comment (keryell) [inline sycl/test/on-device/xocc/simple_tests/matmul.cpp]: While you are here...
```suggestion
    cgh.single_task<class S>([=] {
```

File: sycl/test/on-device/xocc/simple_tests/variant.cpp [hunk 4]
```diff
@@ -0,0 +1,99 @@
...
+  return visit_single_impl<ret_type>(
+      std::forward<Func>(f),
+      typename variant_trait<
+          std::remove_cv_t<std::remove_reference_t<Var>>>::indexes {},
+      std::forward<Var>(var));
+}
+} // namespace detail
+
+template <typename Func, typename Var, typename... Rest>
+auto dev_visit(Func&& f, Var&& var, Rest&&... rest) {
```
Comment (keryell) [inline sycl/test/on-device/xocc/simple_tests/variant.cpp]: You need to add a few comments on the various functions and explain why we need them at the first place

File: sycl/test/on-device/xocc/simple_tests/variant.cpp [hunk 5]
```diff
@@ -0,0 +1,99 @@
...
+  sycl::buffer<std::variant<A, B>> In{d.data(), {d.size()}};
+  sycl::buffer<std::variant<B, A>> Out{d.size()};
+  sycl::buffer<int> OutI{1};
+
+  sycl::queue Queue{};
+
+  Queue.submit([&](sycl::handler &cgh) {
+    auto AIn = In.get_access<sycl::access::mode::read>(cgh);
+    auto AOut = Out.get_access<sycl::access::mode::write>(cgh);
+    cgh.single_task<class Kernel>([=]() {
```
Comment (keryell) [inline sycl/test/on-device/xocc/simple_tests/variant.cpp]:
```suggestion
    cgh.single_task<class Kernel>([=] {
```

**Review by keryell — CHANGES_REQUESTED** [review]
Few details

**Review by keryell — APPROVED** [review]

**Comment by keryell** [comment]
Thanks!

**Pull request merged**